src/ivc/instances_accumulator_computation.rs

```diff
@@ -0,0 +1,226 @@
+/// Module name acronym `instances_accumulator_computation` -> `inst_acc_comp`
+use std::{iter, num::NonZeroUsize};
+
+use halo2_proofs::{
+    halo2curves::ff::{FromUniformBytes, PrimeField, PrimeFieldBits},
+    plonk::Error as Halo2PlonkError,
+};
+use tracing::{debug, instrument};
+
+use crate::{
+    main_gate::{AssignedValue, MainGateConfig, RegionCtx},
+    poseidon::{
+        poseidon_circuit::PoseidonChip, random_oracle::ROCircuitTrait, PoseidonHash, ROTrait, Spec,
+    },
+    util,
+};
+
+/// Default value for use in [`Spec`] in this module
+pub const T: usize = 5;
+
+/// Default value for use in [`Spec`] in this module
+pub const RATE: usize = T - 1;
+
+/// Default value for use in [`Spec`] in this module
+pub const R_F: usize = 10;
+
+/// Default value for use in [`Spec`] in this module
+pub const R_P: usize = 10;
+
+fn default_spec<F: FromUniformBytes<64>>() -> Spec<F, T, RATE> {
+    Spec::new(R_F, R_P)
+}
+
+#[instrument(skip_all)]
+pub fn fold_step_circuit_instances_hash_accumulator<F1, F2>(
+    instances_hash_accumulator: &F1,
+    instances: &[Vec<F1>],
+) -> F1
+where
+    F1: PrimeFieldBits + FromUniformBytes<64>,
```

Comment:
Is there any particular use-case that will use F1!=F2 in this case?
I saw in `fold_assign_step_circuit_instances_hash_accumulator` there is only one Field used.

Reply:
It's always a different field

You don't forget that we have paired circuits and we always fold plonk in different fields
It's just that in the on-circuit version the conversion from one field to another takes place at the `RelaxedPlonkInstance -> AssignedRelaxedPlonkInstance` stage.

Reply:
FIY 
In my next PR, I'll change generic-types to one `C` with internal usage of C::Scalar & C::Base as F1 & F2.

Reply:
Yeah, I remember. My memory is like this: the poseidon hash is from one field to the same field. Then we do the conversion to different field in the next step. And in your approach, you combined these two steps together.